targets/psa_switch/psa_switch.cpp

```diff
@@ -354,31 +372,42 @@ PsaSwitch::ingress_thread() {
     Deparser *deparser = this->get_deparser("ingress_deparser");
     deparser->deparse(packet.get());
 
-    // handling multicast
-    unsigned int mgid = 0u;
-    const auto &f_mgid = phv->get_field("psa_ingress_output_metadata.multicast_group");
-    mgid = f_mgid.get_uint();
-
+    // multicast
+    unsigned int mgid = phv->get_field("psa_ingress_output_metadata.multicast_group").get_uint();
     if(mgid != 0){
-      BMLOG_DEBUG_PKT(*packet,
-                      "Multicast requested for packet with multicast group {}",
-                      mgid);
-      const auto pre_out = pre->replicate({mgid});
-      auto packet_size = packet->get_register(PACKET_LENGTH_REG_IDX);
-      for(const auto &out : pre_out){
-        auto egress_port = out.egress_port;
-        BMLOG_DEBUG_PKT(*packet, "Replicating packet on port {}", egress_port);
-        std::unique_ptr<Packet> packet_copy = packet->clone_with_phv_ptr();
-        packet_copy->set_register(PACKET_LENGTH_REG_IDX, packet_size);
-        enqueue(egress_port, std::move(packet_copy));
-      }
+      multicast(packet.get(), mgid);
       continue;
     }
+    // TODO(pete) from psa spec - else if multicast group is empty, drop packet
+
+    // i2e clone
+    unsigned int clone = phv->get_field("psa_ingress_output_metadata.clone").get_uint();
+    unsigned int clone_session_id = phv->get_field("psa_ingress_output_metadata.clone_session_id").get_uint();
+    if (clone && clone_session_id) {
+      BMLOG_DEBUG_PKT(*packet, "Cloning packet");
+
+      const Packet::buffer_state_t packet_out_state = packet->save_buffer_state();
+      packet->restore_buffer_state(packet_in_state);
+
+      std::unique_ptr<Packet> packet_copy = packet->clone_no_phv_ptr();
+      packet_copy->set_register(PACKET_LENGTH_REG_IDX, ingress_packet_size);
 
-    const auto &f_egress_port = phv->get_field("psa_ingress_output_metadata.egress_port");
-    port_t egress_port = f_egress_port.get_uint();
-    BMLOG_DEBUG_PKT(*packet, "Egress port is {}", egress_port);
+      PHV *copy_phv;
+      copy_phv = packet_copy->get_phv();
+      copy_phv->reset_metadata();
+      copy_phv->get_field("psa_egress_parser_input_metadata.packet_path")
+        .set(PACKET_PATH_CLONE_I2E);
+      copy_phv->get_field("psa_egress_input_metadata.packet_path")
+        .set(PACKET_PATH_CLONE_I2E);
 
+      BMLOG_DEBUG_PKT(*packet, "Cloning packet to CSID {}", clone_session_id);
+      multicast(packet_copy.get(), clone_session_id);
```

Comment:
what is this? That seems very wrong, I don't know how I missed it. Why is the `clone_session_id` used as an mgid here?

I recommend taking a look at the simple_switch implementation again. BTW, in order for P4Runtime to work correctly (but I think that's out-of-scope for now), a cloning session can either point to an egress port or to a multicast group id. Since we are not testing psa_switch with P4Runtime right now, you can ignore the multicast group id case from the simple_switch implementation.

But regardless, multicast & cloning are two different things, and you cannot call `multicast` here.

Reply:
Thanks for clarifying - I'll take a look at this a bit later today.